src/data/Wfin.lean

```diff
@@ -0,0 +1,199 @@
+/-
+Copyright (c) 2019 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jeremy Avigad
+-/
+
+import .equiv.encodable .equiv.list .fin .finset data.fintype
+
+/-!
+# finitely branching W types
+
+Given `α : Type` and `β : α → Type`, the `W` type determined by this data is the
+inductively defined type of trees whose nodes are labeled by elements of `α`,
+where the children of a node labeled `a` are indexed by elements of `β a`.
+
+In this file, we define the special case `Wfin` of finitely branching `W` types,
+where `β a` is equal to `fin (ar a)` for some arity function `ar : α → ℕ`.
+
+The main result in this file is that any such types is encodable if `α` is
+encodable. This can be used to show the encodability of other inductives types,
+such as those that are commonly used to formalize syntax, i.e. terms and
+expressions in a given language. The strategy is illustrated in the example
+found in the namespace `Wfin_example` below. There we define a type of
+propositional formulas over atoms from an encodable type, and show that a map
+from that type to a particular `Wfin` type has a left inverse.
+-/
+
+/--
+`Wfin α ar` is the type of finitely branching trees with labels from `α`, where
+a node labeled `a` has `ar a` children.
+-/
+inductive Wfin {α : Type*} (ar : α → ℕ)
```

Comment:
This could be just `W` defined in the usual way
```lean
inductive W {α : Type*} (β : α → Type*) 
| mk (a : α) (b : β a → W) : W
```
And then include the assumption `[Π a, fintype (β a)]` on all the lemmas about countability